components/tikv_util/src/file.rs

```diff
@@ -134,6 +137,232 @@ impl<R: Read> Read for Sha256Reader<R> {
+/// An instance of `IORateLimiter` should be safely shared between threads.
+#[derive(Debug)]
+pub struct IORateLimiter {
+    unit: atomic::AtomicUsize,
+}
+
+impl IORateLimiter {
+    pub fn new(_unit: usize) -> IORateLimiter {
+        IORateLimiter {
+            unit: atomic::AtomicUsize::new(_unit),
+        }
+    }
+
+    /// Request for token for bytes and potentially update statistics. If this
+    /// request can not be satisfied, the call is blocked. Granted token can be
+    /// less than the requested amount.
+    pub fn request(&self, _io_type: IOType, _io_op: IOOp, bytes: usize) -> usize {
+        std::cmp::min(self.unit.load(atomic::Ordering::Relaxed), bytes)
+    }
+
+    pub fn async_request(
+        &self,
+        io_type: IOType,
+        io_op: IOOp,
+        bytes: usize,
+    ) -> impl Future<Output = usize> {
+        future::ready(self.request(io_type, io_op, bytes))
+    }
+
+    pub fn disable_rate_limit(&self, _io_type: IOType) {}
+
+    pub fn enable_rate_limit(&self, _io_type: IOType) {}
+}
+
+lazy_static! {
+    static ref IO_RATE_LIMITER: Mutex<Option<Arc<IORateLimiter>>> = Mutex::new(None);
+}
+
+pub fn set_io_rate_limiter(limiter: IORateLimiter) {
+    *IO_RATE_LIMITER.lock().unwrap() = Some(Arc::new(limiter));
+}
+
+pub fn get_io_rate_limiter() -> Option<Arc<IORateLimiter>> {
+    if let Some(ref limiter) = *IO_RATE_LIMITER.lock().unwrap() {
+        Some(limiter.clone())
+    } else {
+        None
+    }
+}
+
+/// A wrapper around `std::fs::File` with capability to track and regulate IO flow.
+pub struct InstrumentedFile {
+    inner: fs::File,
+    limiter: Option<Arc<IORateLimiter>>,
+}
+
+impl Deref for InstrumentedFile {
+    type Target = fs::File;
+
+    fn deref(&self) -> &fs::File {
+        &self.inner
+    }
+}
+
+impl DerefMut for InstrumentedFile {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.inner
+    }
+}
+
+impl InstrumentedFile {
+    pub fn open<P: AsRef<Path>>(path: P) -> io::Result<InstrumentedFile> {
+        let inner = fs::File::open(path)?;
+        Ok(InstrumentedFile {
+            inner,
+            limiter: get_io_rate_limiter(),
+        })
+    }
+
+    pub fn create<P: AsRef<Path>>(path: P) -> io::Result<InstrumentedFile> {
+        let inner = fs::File::create(path)?;
+        Ok(InstrumentedFile {
+            inner,
+            limiter: get_io_rate_limiter(),
+        })
+    }
+
+    pub fn try_clone(&self) -> io::Result<InstrumentedFile> {
+        let inner = self.inner.try_clone()?;
+        Ok(InstrumentedFile {
+            inner,
+            limiter: get_io_rate_limiter(),
+        })
+    }
+}
+
+impl Read for InstrumentedFile {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        if let Some(limiter) = &mut self.limiter {
+            let mut remains = buf.len();
+            let mut pos = 0;
+            while remains > 0 {
+                let allowed = limiter.request(get_io_type(), IOOp::Read, remains);
```

Comment:
shall we cap the size of max request size to, say, 1M? same for `write()`.

Reply:
I think that can be for limiter to decide?

Reply:
but `request` is supposed to be a blocking call. say we request 1G write with one `request` call, we wont get a smooth rate limited IO flow, but instead a spike of write after `request` returns.

Reply:
That's not gonna happen, notice that limiter now returns actual granted bytes which can be less than requested bytes, 1G request can be truncated by limiter depending on its strategy, just like RocksDB's design.

Reply:
I see.